### PR TITLE
[Models-CI] Qwen3-VL: migrate to transformers 5.10.2 + move to tier-2 models pipelines

### PR DESCRIPTION
**Pull request by mtairum**
## Summary
Unpins Qwen3-VL from `transformers==4.57.0` (uses the repo-wide 5.10.2), fixes the 5.x incompatibilities it surfaced, and moves its CI from the t3000 pipelines into the tier-based `models_{unit,e2e}_tests` pipelines as **tier 2**. Model: **Qwen/Qwen3-VL-32B-Instruct** (`qwen3-vl-32b`).

## 5.x migration fixes
- **Vision tower** moved `model.visual` → `model.model.visual` (`Qwen3VLModel`) — version-tolerant accessors in `tt/model_config.py`, `demo/demo.py`, `demo/combined.py`, `tt/generator_vllm.py`.
- **Reference dtype**: load the reference vision model in float32 (5.x defaults to the config bf16) so it matches the float32 test inputs.
- **Vision output**: 5.x returns `BaseModelOutputWithDeepstackFeatures`; use `pooler_output` (merged embeds) + `deepstack_features` instead of tuple-unpack / `last_hidden_state` (`test_model`, `test_wrapped_model`).

## Pipeline move (tier 2)
- Added **Qwen3-VL-32B unit** + **e2e** entries to `models_unit_tests.yaml` / `models_e2e_tests.yaml` (commands copied verbatim from the t3000 unit/demo), listed under Tier 2 in `model_ci_tiers.md`.
- Removed `t3k_qwen3_vl` entries from `t3k_unit_tests.yaml` / `t3k_demo_tests.yaml` / `release_tests.yaml` and the orphaned `run_t3000_qwen3_vl_*` shell functions. **vLLM nightly unchanged.**

## Decode-quality fix (#45166 / #48037 regression)
Bringing qwen3-vl onto 5.x surfaced a **BERTScore F1 collapse** (last-good 0.79 → 0.345) and **batch-32 greedy-decode gibberish** in the e2e demo. Root cause is shared with qwen2.5-vl (#47822 / #48037):

- With `allow_force_argmax` disabled (the non-Galaxy default), greedy decode (temp=0 → `k=1,p=0`) goes through the **heavy top-k/top-p multi-all-gather** sampling pipeline, which **corrupts at batch-32 on T3000** — the simple single-gather argmax path is correct.
- Compounded by #45166's separate sampling trace: the force-argmax all-gather's `multi_device_global_semaphore` is frozen into the captured sampling trace, so replay reuses a stale semaphore.

**Fix (mirrors qwen2.5-vl #47822, keeps sampling on-device):**
1. **Enable `allow_force_argmax`** for the qwen3-vl `Transformer` (set in `SAMPLING_AG_CONFIG` before the sampling module is built) → greedy uses the single-gather argmax path.
2. **`_tt_disable_sampling_trace = True`** → run the sampling op eagerly so the all-gather re-acquires a fresh semaphore each step (the heavy decode forward stays traced).
3. **`_tt_vllm_always_refresh_decode_trace_inputs = True`** → re-stage decode trace inputs from host each step (also fixes the #45522 repeat-batch staleness).
4. Shared `tt_transformers/tt/generator.py` honors `_tt_disable_sampling_trace` in `sample_decode_on_device` (identical to #47822; dedupes on rebase).
5. Demo keeps on-device sampling by default with a `TT_QWEN_FORCE_HOST_SAMPLING=1` A/B toggle.

This keeps decode **on-device** (no per-step logits read-back), so accuracy *and* perf are preserved.

## Validation
- `t3k_qwen3_vl_tests` unit suite **green** (all 9 tests pass).
- **e2e demo green** (on-device force-argmax): BERTScore F1 **0.788**, both `bert-score` + `text-only` pass, decode **~18.9 tok/s/user** (vs ~13.3 with the earlier host-argmax approach — ~42% faster) — [run 28233883882](https://github.com/tenstorrent/tt-metal/actions/runs/28233883882). Previously a known failure on main (4.57.0); now passes.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:mtairum/unpin-qwen3vl)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:mtairum/unpin-qwen3vl)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:mtairum/unpin-qwen3vl)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:mtairum/unpin-qwen3vl)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:mtairum/unpin-qwen3vl)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:mtairum/unpin-qwen3vl)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:mtairum/unpin-qwen3vl)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:mtairum/unpin-qwen3vl)

### CODE REVIEW

**Comment by mtairum** [comment]
## 🔬 Investigating the `bert-score` e2e failure (BERTScore F1)

### Status
- ✅ unit tests green; ✅ `text-only` e2e green (the 5.x model crashes are fixed: `get_rope_index` `mm_token_type_ids`, rotary `torch.device`, bert_score `model_max_length` clamp).
- ❌ `bert-score` e2e fails on **BERTScore F1 ≈ 0.345** (threshold 0.70). Branch rebased onto latest main.

### Root cause (this is a pre-existing main bug, not introduced by this PR)
The model produces **degenerate greedy generation** (e.g. `"This image is a a"` → token salad) instead of a real description. Confirmed:
- **Bisected on main:** last good F1 **0.790** (06-05, run 26988595777) → **0.667** (06-08) → **0.345** (06-12 onward). Two drops; the severe one lands in the **06-10→06-12** window.
- **Prime suspect: #45166** "Clean up sampling separation in generators" — it restructured the on-device **sampled-token feedback** in `tt_transformers/tt/generator.py`, which qwen3-vl inherits (`qwen3_vl.tt.generator.Generator` → `TTTGenerator`).
- Parity check: ours **0.348** vs main **0.345** — i.e. this PR is a strict improvement (text-only fixed) at parity with main; the F1 gap exists on main too.

### What we're trying
1. **Confirmed partial lever:** setting gemma4's `_tt_vllm_always_refresh_decode_trace_inputs = True` on qwen3-vl's `Transformer` recovered F1 **0.348 → 0.552** in an earlier experiment. Rationale: qwen3-vl's on-device decode-trace token buffer (`device_inputs[0]`) isn't a valid rank-4 `ttnn.sampling` output, so feeding the sampled token back corrupts decode (same carve-out gemma4 needed for #45166).
2. **Current experiment** (branch `mtairum/qwen3vl-decode-fix-dbg`, run [28163200810](https://github.com/tenstorrent/tt-metal/actions/runs/28163200810)): apply that flag on the rebased branch + observe the residual generated text to find what's still missing (target F1 > 0.70).

### Next steps
- If the flag alone doesn't clear 0.70, instrument the decode to dump generated tokens vs a torch-HF greedy reference and pinpoint the residual (likely the second, milder 0.790→0.667 commit and/or incomplete decode-trace staging).
- Fix will likely land as a separate "qwen3-vl decode fix" once it clears 0.70 (it's a main-bug fix orthogonal to this unpin).

**Comment by mtairum** [comment]
### Decode-fix experiment result: flag confirmed (partial), residual is sporadic token corruption

Applied gemma4's `_tt_vllm_always_refresh_decode_trace_inputs = True` to qwen3-vl's `Transformer` on the rebased branch (`mtairum/qwen3vl-decode-fix-dbg`) and re-ran the bert-score e2e ([run 28163200810](https://github.com/tenstorrent/tt-metal/actions/runs/28163200810)).

**Result: F1 0.345 → 0.549** (reproduces the earlier 0.552). Big qualitative improvement — output went from `"This image is a a"` to long, mostly-coherent descriptions, e.g.:
> "This image is a vibrant, highly saturated, … composed landscape photograph that evokes a sense of … tranquility, and idealized nature. Here's a detailed description:"

…but with **sporadic wrong tokens injected** (`上级ly`, `ackers`, `的身体ically`, `צרה`), one sample degenerating into token salad. So:
- The flag fixes the #45166 **token-feedback buffer corruption** (necessary).
- A **residual decode-precision issue** remains: greedy (temp=0) decode occasionally picks a wrong token. F1 0.549 is still < 0.70.

Recovery math: last-good was **0.790** (06-05); two regressions (mild 0.790→0.667, severe →0.345). The flag recovers to 0.549 — still below even the post-mild 0.667, so there's both residual severe + the mild-drop commit to close.

**Next step:** instrument the decode to compare TT generated token IDs against a torch-HF greedy reference (same prompt) to localize the first divergence — is it a near-tie argmax (precision) or structural? Plus bisect the separate mild-drop commit (06-05→06-08).

**Comment by mtairum** [comment]
### Diagnostic update — accuracy-precision run invalid; pivoting to host-argmax

**Accuracy-precision experiment (run 28175972143): no signal.** Setting `DecodersPrecision.accuracy` (to test whether the residual F1~0.55 is decode precision) needs a tensor cache for an uncached dtype, but the MLPerf cache mount is **read-only** → the model never ran:
```
TT_FATAL: Cannot open ".../tensor_cache_instruct_bfp8/layers.0.attention.wqkv...tensorbin"
for writing: errno=30 "Read-only file system"
```
So the precision hypothesis can't be tested this way without write access to the cache.

**Pivot — force HOST argmax (run 28200327150).** Higher-value test anyway: at temp=0 the demo samples **on-device** (`argmax_on_device=True`, `out_tok = logits.unsqueeze(1)`), and the #45166 regression was exactly *generator sampling-separation* (moving sampling on-device). So the on-device decode sampler is the prime suspect for the sporadic wrong tokens. This experiment sets `argmax_on_device=False` → `decode_forward` returns full logits → `sample_host` does a host fp32 argmax, bypassing the on-device sampler (keeps `.performance` precision, so it reuses the existing cache — no read-only-FS issue). If F1 recovers from ~0.55, the on-device sampler is the culprit.

**Comment by mtairum** [comment]
## ✅ Root cause pinned — F1 fully recovered (0.791)

The **force-host-argmax** experiment (run [28200327150](https://github.com/tenstorrent/tt-metal/actions/runs/28200327150)) recovered decode quality completely:
- **Mean BERTScore F1: 0.791** (last-good was 0.790; threshold 0.70) — **2 passed** (bert-score + text-only)
- Generated text is clean, no sporadic garbage tokens:
  > *"This image is a vibrant, highly saturated, and symmetrical landscape photograph that evokes a sense of serene, idyllic beauty. It features a picturesque rural scene with a striking mirror-like reflection..."*

### The two regressions, both now explained
1. **#45166 trace-input feedback staleness** → fixed by the gemma4-style `_tt_vllm_always_refresh_decode_trace_inputs=True` on the qwen3-vl `Transformer` (recovered 0.345 → 0.549).
2. **#45166 on-device decode sampler** → forcing **host fp32 argmax** (`argmax_on_device=False` → `decode_forward` returns full logits → `sample_host` argmax) recovered 0.549 → **0.791**. The on-device greedy sampler introduced in the sampling-separation refactor was selecting wrong tokens (the sporadic garbage); host argmax over the same logits is correct.

So the residual was **not** a model-precision issue — it was the on-device sampler. The current experiment branch (flag + host-argmax) is green at F1 0.791.

### Productionization — open question
Forcing host argmax restores correctness but moves per-step logits to host (a decode-throughput cost), and it only addresses the demo path — the same on-device sampler is shared via `tt_transformers` and likely affects production/vLLM serving too. Options: (a) scoped host-sampling fix for qwen3-vl now + issue for the shared on-device sampler; (b) fix the shared on-device decode sampler directly. Leaning (a) to unblock, with a tracked follow-up.

**Comment by mtairum** [comment]
✅ **Post-rebase validation green — ready for review.** Rebased onto latest main (includes the decode-path changes #48042 / #46462), e2e re-validated: **Mean BERTScore F1 0.789**, both `bert-score` + `text-only` demos pass — [run 28229825203](https://github.com/tenstorrent/tt-metal/actions/runs/28229825203). Decode fix holds on the current base. Marking ready.

File: models/demos/qwen3_vl/tt/model_config.py [hunk 1]
```diff
@@ -105,15 +105,20 @@ def get_state_dict_prefix(self, module_name, layer_num=None, deepstack_merger_nu
...
         config = AutoModelForCausalLM.config_class.from_pretrained(self.CKPT_DIR)
         config.vision_config.depth = depth if depth is not None else config.vision_config.depth
-        model = AutoModelForCausalLM.from_pretrained(self.CKPT_DIR, config=config)
-        return model.visual
+        # transformers 5.x loads in the config dtype (bf16) by default; force float32 so the reference
+        # and its sub-modules match the float32 test inputs (4.x defaulted to float32).
+        model = AutoModelForCausalLM.from_pretrained(self.CKPT_DIR, config=config, torch_dtype=torch.float32)
+        # transformers 5.x nests the vision tower under model.model.visual (Qwen3VLModel); 4.x exposed
+        # it at the top level as model.visual.
+        return model.visual if hasattr(model, "visual") else model.model.visual
```
[the remark this answers is not in the document]
Comment (mtairum) [reply models/demos/qwen3_vl/tt/model_config.py]: Done in 95ea80d — extracted `get_hf_visual(model)` into `tt/common.py` and replaced all 5 call sites (model_config, generator_vllm, demo ×2, combined), dropping the duplicated comments. Thanks!

**Comment by mtairum** [comment]
Addressed the review bot comments in 95ea80d:
- **Copilot** (duplicated vision-tower accessor across 5 sites): extracted `get_hf_visual(model)` into `tt/common.py` and call it from `model_config.py`, `generator_vllm.py`, `demo.py` (×2), and `combined.py` — single source of truth, duplicated comments removed.
- **pre-commit / reviewdog** (demo.py bert_score imports): reordered the local imports (`import` before `from`); isort check is clean.

Pure refactor + import-order — no behavior change. Decode fix unaffected (still F1 0.789).

**Review by yieldthought — APPROVED** [review]

**Review by roseli-TT — APPROVED** [review]

**Review by uaydonat — APPROVED** [review]

**Review by gwangTT — APPROVED** [review]
